source/simulator/assemblers/stokes.cc

```diff
@@ -0,0 +1,393 @@
+  local_assemble_stokes_mass_reference_density (const double                                     pressure_scaling,
+                                                const bool                                       /*rebuild_stokes_matrix*/,
+                                                internal::Assembly::Scratch::StokesSystem<dim>  &scratch,
+                                                internal::Assembly::CopyData::StokesSystem<dim> &data,
+                                                const Parameters<dim> &parameters) const
+  {
+    // assemble RHS of:
+    //  - div u = 1/rho * drho/dz g/||g||* u
+    Assert(parameters.formulation_mass_conservation ==
+           Parameters<dim>::FormulationMassConservation::reference_density_profile,
+           ExcInternalError());
+
+    const Introspection<dim> &introspection = this->introspection();
+    const FiniteElement<dim> &fe = this->get_fe();
+    const unsigned int stokes_dofs_per_cell = data.local_dof_indices.size();
+    const unsigned int n_q_points    = scratch.finite_element_values.n_quadrature_points;
+
+    for (unsigned int q=0; q<n_q_points; ++q)
+      {
+        for (unsigned int i=0, i_stokes=0; i_stokes<stokes_dofs_per_cell; /*increment at end of loop*/)
+          {
+            if (introspection.is_stokes_component(fe.system_to_component_index(i).first))
+              {
+                scratch.phi_p[i_stokes] = scratch.finite_element_values[introspection.extractors.pressure].value (i, q);
+                ++i_stokes;
+              }
+            ++i;
+          }
+
+        const Tensor<1,dim>
+        gravity = this->get_gravity_model().gravity_vector (scratch.finite_element_values.quadrature_point(q));
+        const double drho_dz_u = scratch.adiabatic_density_gradients[q]
+                                 * (gravity * scratch.velocity_values[q]) / gravity.norm();
+        const double one_over_rho = 1.0/scratch.mass_densities[q];
+        const double JxW = scratch.finite_element_values.JxW(q);
+
+        for (unsigned int i=0; i<stokes_dofs_per_cell; ++i)
+          data.local_rhs(i) += (pressure_scaling *
+                                one_over_rho * drho_dz_u * scratch.phi_p[i])
+                               * JxW;
+      }
+  }
+
+  template <int dim>
+  void
+  StokesAssembler<dim>::
+  local_assemble_stokes_mass_implicit_reference_density (const double                                     pressure_scaling,
+                                                         const bool                                       rebuild_stokes_matrix,
+                                                         internal::Assembly::Scratch::StokesSystem<dim>  &scratch,
+                                                         internal::Assembly::CopyData::StokesSystem<dim> &data,
+                                                         const Parameters<dim> &parameters) const
+  {
+    // assemble compressibility term of:
+    //  - div u - 1/rho * drho/dz g/||g||* u = 0
```

Comment:
It works correctly. I have two tests, one that uses the explicit form and one that uses this implicit form. The latter convergence in one Newton iteration, so I am pretty sure it works. As to the linear solver: yes, it might be worse and it needs more testing. If you look at  ``tangurnis_tala[_implicit]/screen_output``, they both need 30+2 iterations.